### ISSUE TITLE
kube-proxy pods are not getting deleted after grace period

### ISSUE THREAD

**Issue by gmarek** [kubernetes/kubernetes#18429]
I have an ordinary GCE e2e cluster running on flannel (NETWORK_PROVIDER=“flannel”). After starting the 500 Node cluster some kube-proxy pods are pending (I've seen 1 and 3) with "ErrImagePull" errors. I tried to delete them normally (`kubectl --namespace=kube-system delete pod kube-proxy-blablabla`)
```
kube-proxy-blablabla              0/1       Terminating   0          42m
```
and waited for 10 minutes.
```
kube-proxy-blablabla              0/1       Terminating   0          54m
```

In the end I had to force deletion (`kubectl --namespace=kube-system delete pod kube-proxy-blablabla --grace-period=0`) which worked.

cc @wojtek-t @smarterclayton @bprashanth @kubernetes/goog-node 

Assigning to @dchen1107 to reassing.

**Comment by yujuhong** [comment]
kube-proxy has been converted to a static pod (#17121). You need to delete the manifest file itself (`/etc/kubernetes/manifests/kube-proxy.manifest`) to truly remove the pod. Deleting the mirror pod at the apiserver has no effect on the static pod.

We used to recreate the mirror pod when it diverges from the spec of the static pod. This is no longer the case since we only compare the annotation now to avoid problems caused by version skews (#15960). Unfortunately, we currently don't have any mechanism to prevent users from modifying the mirror pods. More discussions are in #16627.

**Comment by gmarek** [comment]
I don't want to delete the pod, I want to force a retry. After pod got deleted it was recreated and it was working perfectly.

**Comment by smarterclayton** [comment]
kubectl delete --grace-period=0 - I suspect that the kubelet stomped your
delete with its own logic (it regards the mirror cache as its property).  I
think the mirror pod should respond to restarts that way though.

**Comment by yujuhong** [comment]
> I don't want to delete the pod, I want to force a retry. After pod got deleted it was recreated and it was working perfectly.

kubelet would automatically retries if there is any error. Deleting the mirror pod doesn't really help in this case, IMO.

**Comment by yujuhong** [comment]
> kubectl delete --grace-period=0 - I suspect that the kubelet stomped your
> delete with its own logic (it regards the mirror cache as its property).  I
> think the mirror pod should respond to restarts that way though.

@smarterclayton, kubelet doesn't respond to changes made to the mirror pods by the users, so it would not try to restart containers because of that.

**Comment by yujuhong** [comment]
@gmarek, the ErrImagePull error might be related to a recent bug causing stale status for static pods. My fix has just been merged (#18129). Let me know if you see it again (or just file a bug).

As for mirror pods, it would not be correctly deleted unless you force `-grace-period=0`. It's not really a bug as users are not supposed to modify the mirror pods. Closing this bug for now.
